src/Microsoft.ML.SEAL/EncryptedLinearModelParameters.cs

```diff
@@ -0,0 +1,604 @@
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.ML.Calibrators;
+using Microsoft.ML.Data;
+using Microsoft.ML.Internal.Utilities;
+using Microsoft.ML.Model;
+using Microsoft.ML.Model.OnnxConverter;
+using Microsoft.ML.Model.Pfa;
+using Microsoft.ML.Numeric;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Transforms;
+using Microsoft.Research.SEAL;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.ML.Trainers
+{
+    /// <summary>
+    /// Base class for linear model parameters.
+    /// </summary>
+    public abstract class EncryptedLinearModelParameters : ModelParametersBase<float>,
+        IValueMapperEncrypted,
+        ICanSaveInIniFormat,
+        ICanSaveInTextFormat,
+        ICanSaveInSourceCode,
+        ICanGetSummaryAsIRow,
+        ICanSaveSummary,
+        IPredictorWithFeatureWeights<float>,
+        IFeatureContributionMapper,
+        ICalculateFeatureContribution,
+        ISingleCanSavePfa,
+        ISingleCanSaveOnnx
+    {
+        [BestFriend]
+        private protected readonly VBuffer<float> Weight;
+
+        // _weightsDense is not persisted and is used for performance when the input instance is sparse.
+        private VBuffer<float> _weightsDense;
+        private readonly object _weightsDenseLock;
+
+        private sealed class WeightsCollection : IReadOnlyList<float>
+        {
+            private readonly EncryptedLinearModelParameters _pred;
+
+            public int Count => _pred.Weight.Length;
+
+            public float this[int index]
+            {
+                get
+                {
+                    Contracts.CheckParam(0 <= index && index < Count, nameof(index), "Out of range");
+                    float value = 0;
+                    _pred.Weight.GetItemOrDefault(index, ref value);
+                    return value;
+                }
+            }
+
+            public WeightsCollection(EncryptedLinearModelParameters pred)
+            {
+                Contracts.AssertValue(pred);
+                _pred = pred;
+            }
+
+            public IEnumerator<float> GetEnumerator()
+            {
+                return _pred.Weight.Items(all: true).Select(iv => iv.Value).GetEnumerator();
+            }
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        /// <summary> The predictor's feature weight coefficients.</summary>
+        public IReadOnlyList<float> Weights => new WeightsCollection(this);
+
+        /// <summary> The predictor's bias term.</summary>
+        public float Bias { get; protected set; }
+
+        private List<Plaintext> _encodedWeights;
+
+        private Evaluator _evaluator;
+
+        private CKKSEncoder _ckksEncoder;
+
+        private readonly DataViewType _inputType;
+
+        bool ICanSavePfa.CanSavePfa => true;
+
+        bool ICanSaveOnnx.CanSaveOnnx(OnnxContext ctx) => true;
```

Comment:
>CanSaveOnnx [](start = 26, length = 11)

are you sure this model can also be saved as ONNX?

Reply:
There is certainly no reason it can't be saved as a standard SDCA ONNX model. The key to keep in mind about this change is that it only affects evaluation, so all the data is unencrypted and shouldn't have any issue being written out as an ONNX model. Is there something specific you are concerned over?